### ISSUE TITLE
Duplicate nodes when using reactor's subscribe method.

### ISSUE THREAD

**Issue by Sm0keySa1m0n** [spring-projects/spring-data-neo4j#2247]
When saving two nodes with the same ID by calling `ReactiveNeo4jRepository::save` and subscribing to the result, two duplicate nodes are saved to the database (even when using optimistic locking, they get assigned the same version). When calling the block method on `Mono::save`, it saves without duplicating the node. Therefore, this seems to be a concurrency issue as it only occurs when saving on two different threads (which shouldn't be the case if using optimistic locking).

This issue occurs on 6.0.8 and 6.1.1-SNAPSHOT (at the time of writing this). To reproduce, create a very simple entity class with just an ID field (I tested this with both a string and long ID type) and also a version field. I then created two instances of the entity class with the same ID and then saved each of them, subscribing to the result. You'll observe two nodes are persisted to Neo4j both with the same ID and the same version. When calling block on the save Mono, this issue does not occur.

**Comment by Sm0keySa1m0n** [comment]
If you use a standard imperative `Neo4jRepository` and wrap the save call in a Mono sink, and repeat the test I stated above, an appropriate `OptimisticLockingFailureException` is thrown.

**Comment by Sm0keySa1m0n** [comment]
The issue also occurs using an imperative `Neo4jRepository` if calling save from two different threads.

**Comment by meistermeier** [comment]
Thanks for taking the time and reporting this. I will have a look but this behaviour seems legit for a certain scenario.
Given that both transactions from different threads take place in parallel in the database (which still means that there is an order when the TXs get committed), the creation code in SDN will look if a node with the same id already exists. This is not the case because transaction visibility will hide the other node in creation and the nodes will get created.
To prevent such situations, you should manually create a uniqueness constraint on the custom id field. Otherwise the database itself treats the custom id field as a simple property. With a constraint, the logical second transaction will fail.

**Comment by Sm0keySa1m0n** [comment]
What do you mean by manually create a uniqueness constraint?

**Comment by Sm0keySa1m0n** [comment]
Never mind, I assume you're referring to this: https://neo4j.com/docs/cypher-manual/current/administration/constraints/

**Comment by meistermeier** [comment]
Using the version / optimistic locking with an externally provided id is something I would always advise. Otherwise those entities will always be treated as new and e.g. relationship changes cannot be tracked correctly.
With the `@Version` annotated property in place, it will always check if a node with the id and version exist in the database. And only created a new one if this is not the case.
If in a multi-threading situation the creation would fail of the "last new node", a database exception will occur and you would have to either
- retry with load, manipulate and save (e.g. different property changes)
- retry with load, change the logic because you have now a conflict and this needs to get merged (e.g. some property change)
- dismiss (just the creation of the now already existing node)
 your operation.
You would have to wrap this in a retry logic (like resillience4j or similar) but this should apply to all interactions with external systems in general.

**Comment by Sm0keySa1m0n** [comment]
Ahh, I thought you were saying optimistic locking was the solution. I'll implement that constraint then; unless there's a way of synchronously performing transactions on application side or something to that effect to make them aware of each other before hitting the database?

**Comment by meistermeier** [comment]
No. There used to be one in Neo4j-OGM but there are reasons why to separate those concerns (object mapping and database administration). One of them (and my personal most important one):
If the syntax for constraint (and indexes) changes, we would have not to patch this in but also define a bottom line of version which support a newer version X of Neo4j. This would make SDN less flexible for the database versions you can work with.
When we redesigned SDN we wanted to keep such infrastructure features out of the scope of an object mapping library and as a side-product https://github.com/michael-simons/neo4j-migrations/ were created. Also there is https://github.com/liquibase/liquigraph/ to take care of those things.
